collagraph/cgx/cgx.py

```diff
@@ -54,214 +245,230 @@ class Foo(cg.Component):
     script = parser.root.child_with_tag("script").data
 
-    # Compile for the first time to get a handle on the
-    # context that is defined in the components' script
-    preliminary_component_type, context = load_code(script)
-
-    # Construct a render function from the template block
-    template_root_node = parser.root.child_with_tag("template").children[0]
-    compiled_tree = template_root_node.compile(preliminary_component_type, context)
-
-    def render(self):
-        return compiled_tree.to_vnode(self, context)
-
-    # NOTE: load_code is called multiple times, to people need
-    # to be aware that module-level side-effects will be executed
-    # multiple times. Please put any side-effects in your compontent.mounted()
-    # and don't forget to clean-up in component.before_unmount() if needed.
-    return load_code(script, render_function=render)
-
-
-def load_code(script, render_function=None):
-    """
-    Exec the script with a custom (globals) dict to capture
-    all the defined classes and methods.
-    """
-    context = {}
-    ComponentMeta.RENDER_FUNCTION = render_function
-    try:
-        exec(script, context)
-    finally:
-        ComponentMeta.RENDER_FUNCTION = None
-
-    results = re.search(COMPONENT_CLASS_DEFINITION, script)
-    if not results:
-        raise ValueError(f"Could not find a class definition in:\n{script}")
+    script_tree = ast.parse(script, filename=str(path), mode="exec")
+
+    # Inject 'import collagraph' into imports
+    script_tree.body.insert(
+        0,
+        ast.ImportFrom(
+            module="collagraph",
+            names=[
+                ast.alias(
+                    name="create_element",
+                    asname="_create_element",
+                )
+            ],
+        ),
+    )
 
-    component_type = None
-    for class_name in results.groups():
-        if class_definition := context.get(class_name):
-            if issubclass(class_definition, Component):
-                if component_type is not None:
-                    raise ValueError(
-                        f"Multiple component class definitions found:\n{script}"
-                    )
-                component_type = class_definition
+    script_tree.body.insert(
+        1,
+        ast.parse(
+            # TODO: add self.state and self.props as searchable namespaces
+            # TODO: caching of lookup source
+            textwrap.dedent(
+                """
+                def _lookup(self, name):
+                    if hasattr(self, name):
+                        return getattr(self, name)
+                    if name in globals():
+                        return globals()[name]
+                    raise NameError(f"name '{name}' is not defined")
+                """
+            ),
+            mode="exec",
+        ).body[0],
+    )
 
-    if not component_type:
-        raise ValueError(f"No subclass of Component found:\n{script}")
+    # Find ClassDef
```

Comment:
ultimately I guess you want to make this just as smart as the code you had behavior, to identify a subclass of Component, and not just any class definition